aws/s3.go

```diff
@@ -304,119 +304,181 @@ func excludeBucketByREList(bucketName string, reList []*regexp.Regexp) bool {
-	identifiers := []*s3.ObjectIdentifier{}
+// emptyBucket will empty the given S3 bucket by deleting all the objects that are in the bucket. For versioned buckets,
+// this includes all the versions and deletion markers in the bucket.
+// NOTE: In the progress logs, we deliberately do not report how many pages or objects are left. This is because aws
+// does not provide any API for getting the object count, and the only way to do that is to iterate through all the
+// objects. For memory and time efficiency, we opted to delete the objects as we retrieve each page, which means we
+// don't know how many is left until we complete all the operations.
+func emptyBucket(svc *s3.S3, bucketName *string, isVersioned bool, batchSize int) error {
+	// Since the error may happen in the inner function handler for the pager, we need a function scoped variable that
+	// the inner function can set when there is an error.
+	var errOut error
+	errOut = nil
+	pageId := 1
 
 	// Handle versioned buckets.
 	if isVersioned {
 		err := svc.ListObjectVersionsPages(
 			&s3.ListObjectVersionsInput{
-				Bucket: bucketName,
+				Bucket:  bucketName,
+				MaxKeys: aws.Int64(int64(batchSize)),
 			},
 			func(page *s3.ListObjectVersionsOutput, lastPage bool) (shouldContinue bool) {
-				for _, obj := range page.Versions {
-					logging.Logger.Debugf("Bucket %s object %s version %s", *bucketName, *obj.Key, *obj.VersionId)
-					identifiers = append(identifiers, &s3.ObjectIdentifier{
-						Key:       obj.Key,
-						VersionId: obj.VersionId,
-					})
+				logging.Logger.Debugf("Deleting object page %d (%d objects) from bucket %s", pageId, len(page.Versions), aws.StringValue(bucketName))
+				if err := deleteObjectVersions(svc, bucketName, page.Versions); err != nil {
+					logging.Logger.Errorf("Error deleting objects for page %d from bucket %s: %s", pageId, aws.StringValue(bucketName), err)
+					errOut = err
+					return false
 				}
-				for _, obj := range page.DeleteMarkers {
-					logging.Logger.Debugf("Bucket %s object %s DeleteMarker %s", *bucketName, *obj.Key, *obj.VersionId)
-					identifiers = append(identifiers, &s3.ObjectIdentifier{
-						Key:       obj.Key,
-						VersionId: obj.VersionId,
-					})
+				logging.Logger.Infof("[OK] - deleted object page %d (%d objects) from bucket %s", pageId, len(page.Versions), aws.StringValue(bucketName))
+
+				logging.Logger.Debugf("Deleting object page %d (%d deletion markers) from bucket %s", pageId, len(page.DeleteMarkers), aws.StringValue(bucketName))
+				if err := deleteDeletionMarkers(svc, bucketName, page.DeleteMarkers); err != nil {
+					logging.Logger.Errorf("Error deleting deletion markers for page %d from bucket %s: %s", pageId, aws.StringValue(bucketName), err)
+					errOut = err
+					return false
 				}
+				logging.Logger.Infof("[OK] - deleted object page %d (%d deletion markers) from bucket %s", pageId, len(page.DeleteMarkers), aws.StringValue(bucketName))
+
+				pageId++
 				return true
 			},
```

Comment:
The key change is here, where instead of getting all the bucket objects and then calling delete on the combined list, we call the delete operation in the page handler so that we delete each page as it is queried.

I was worried if this would cause a race condition since we are modifying the list that is being queried, but empirical testing shows that AWS handles this correctly (see below for annotation in test highlighting the empirical test).